python/pyspark/sql/group.py

```diff
@@ -194,6 +194,65 @@ def pivot(self, pivot_col, values=None):
             jgd = self._jgd.pivot(pivot_col, values)
         return GroupedData(jgd, self.sql_ctx)
 
+    def apply(self, udf):
+        """
+        Maps each group of the current :class:`DataFrame` using a pandas udf and returns the result
+        as a :class:`DataFrame`.
+
+        The user-function should take a `pandas.DataFrame` and return another `pandas.DataFrame`.
+        Each group is passed as a `pandas.DataFrame` to the user-function and the returned
+        `pandas.DataFrame` are combined as a :class:`DataFrame`. The returned `pandas.DataFrame`
+        can be arbitrary length and its schema should match the returnType of the pandas udf.
+
+        :param udf: A wrapped function returned by `pandas_udf`
+
+        >>> df = spark.createDataFrame(
+        ...     [(1, 1.0), (1, 2.0), (2, 3.0), (2, 5.0), (2, 10.0)],
+        ...     ("id", "v"))
+        >>> @pandas_udf(returnType=df.schema)
+        ... def normalize(pdf):
+        ...     v = pdf.v
+        ...     return pdf.assign(v=(v - v.mean()) / v.std())
+        >>> df.groupby('id').apply(normalize).show() # doctest: + SKIP
+        +---+-------------------+
+        | id|                  v|
+        +---+-------------------+
+        |  1|-0.7071067811865475|
+        |  1| 0.7071067811865475|
+        |  2|-0.8320502943378437|
+        |  2|-0.2773500981126146|
+        |  2| 1.1094003924504583|
+        +---+-------------------+
+
+        .. seealso:: :meth:`pyspark.sql.functions.pandas_udf`
+
+        """
+        from pyspark.sql.functions import pandas_udf
+
+        # Columns are special because hasattr always return True
+        if isinstance(udf, Column) or not hasattr(udf, 'func') or not udf.vectorized:
+            raise ValueError("The argument to apply must be a pandas_udf")
+        if not isinstance(udf.returnType, StructType):
+            raise ValueError("The returnType of the pandas_udf must be a StructType")
+
+        df = DataFrame(self._jgd.df(), self.sql_ctx)
+        func = udf.func
+        returnType = udf.returnType
+
+        # The python executors expects the function to take a list of pd.Series as input
+        # So we to create a wrapper function that turns that to a pd.DataFrame before passing
+        # down to the user function
+        columns = df.columns
+
+        def wrapped(*cols):
+            import pandas as pd
+            return func(pd.concat(cols, axis=1, keys=columns))
```

Comment:
What do you all think about making the `ArrowPandasSerializer` also able to serialize pandas.DataFrames?  Then it wouldn't require this extra wrapping and I think it could be useful for other things in the future as well.  @HyukjinKwon @ueshin @viirya ?

Reply:
I think we need to think a little more about how do we handle different formats of arrow data. 

Currently, the input of arrow serializer is a list of (pd.Series, DataType), I feel it's cleaner that this class not deal with type coercion and just serialization. It could take a `pyarrow.Table` for instance and let caller construct the `pyarrow.Table`.

Another thing to think about is whatever the data we are passing are not purely `pd.Series` and `pd.DataFrame`. What if, for instance, we want to serialize a (pd.Series, pd.DataFrame) tuple or a tuple of (scalar value, pd.DataFrame). Maybe somehow making the serializer composable is more flexiable. i.e. a class knows how to serialize `pd.Series`, a class knows how to serialize `pd.DataFrame` and if we want to serialize (pd.Series, pd.DataFrame) tuple we can compose them.

Reply:
That's fine, we can leave the serializer as is, but it seems like a waste to unwrap the pandas_udf only to then wrap it again as another pandas_udf.  @icexelloss , how about changing the Series to a DataFrame in `worker.py` `wrap_pandas_udf` right before the user function is called?  

That way both transformations are in one place and `wrap_pandas_udf` would work like this:
Changes Series to DataFrame; Call user function with DataFrame; Get result DataFrame; Change DataFrame to Series.

Reply:
@BryanCutler yeah I was trying to do that earlier  that but unfortunately the column names are lost on the worker so we cannot construct the `Pandas.DataFrame` on the worker.

I think the best plcae to define the wrap function is probably on the pyspark driver side because we have the most information there. However, that requires some refactoring. I will give it a try and see how that goes.

Reply:
I see, yes it does strip the names from the columns.  How about the opposite way then, move the conversion from DataFrame to Series out of `worker.py` `wrap_pandas_udf` and put it here?  I think you could then use `wrap_pandas_udf` as-is.

I also think doing this wrapping would be better off in `functions.py` `_create_udf`, is that possible?

Reply:
I was thinking it should simplify things. Mind if I give it a try? If does only complicate then we can forget it..

Reply:
Feel free! The tricky bit is that the wrapped function is unknown when calling pandas_udf so there needs to be a way to change it later.

Also the function chaining makes it a bit complicated too.

I still think we should do it, but needs to be more careful.

Reply:
ok, I'll try but if this PR is ready to merge then no need to hold up for this, we can address later.